### DISCUSSION TITLE
Idea for the roadmap

### DISCUSSION THREAD

**Discussion by hape42** [hape42/dailygammon#76] [Ideas]
If someone has an idea, what the app can be further fixed, then just write a new post here in the discussion. No idea is too crazy not to at least talk about it.

**Comment by b-dubbs** [comment]
Perhaps already in your backlog (though I didn't see it on my scan of your roadmap): match reminder notifications similar to those available by using the DGText2Area extension

https://chrome.google.com/webstore/detail/dgtext2area/falmnkmhomkpphcpkdflemlahipdnemp

The extension offers multiple features, but (IMHO) the most-used function is my ask for your future roadmap:
- You can now set and remove a reminder text for a certain match. Like "Double the cube". These reminders are stored in the browser LocalStorage for the site. One reminder line for a match.
![ChromeExtension](https://user-images.githubusercontent.com/5258161/211154633-6292c0fd-c5ff-4618-9c82-656736fea0ed.jpg)

**Comment by kagsteiner** [comment]
My #1 issue is performance. Probably not much we can to as the app needs to talk to DG all the time, but we should see whether we do any network call to DG twice, what we can do asynchronously, and (big, probably impossible and a complete redesign) keep a local board representation for trying out different moves and only communicate to DG once the player has decided for a move.

**Reply by hape42** [reply]
>  and (big, probably impossible and a complete redesign) keep a local board representation for trying out different moves and only communicate to DG once the player has decided for a move.

Unfortunately, I do not understand this point.
You can already make moves and reset them with "undo" without finally sending the move to the server?

**Reply by hape42** [reply]
> My #1 issue is performance. Probably not much we can to as the app needs to talk to DG all the time, but we should see whether we do any network call to DG twice, what we can do asynchronously ...

I fully agree with you. There is still a lot of room for improvement.

**Reply by kagsteiner** [reply]
Okay, maybe I must be more precise what my observation and problem is :-)

Let's assume I'm in a game. I click "roll dice" and see the dice
* I click on a checker to move it. Unexpectedly much time passes, it's not immediate that the checker moves
* I click on a second checker to move it. Again more time than I know from off-line backgammons passes
* I see that the position is not good and want to try out a different move. I press "undo" and again some time passes until the position before my two checker moves is restored.

Sometimes the time in these 3 steps is really fast, sometimes it is slow, so my assumption is that the DG app does some remote query to the DG server happens whenever I click a checker or undo.

**Reply by kagsteiner** [reply]
And of course, when I just now tried to reproduce my performance issues in the simulator, the app was **blindingly fast** :-)

**Reply by hape42** [reply]
> And of course, when I just now tried to reproduce my performance issues in the simulator, the app was **blindingly fast** :-)

Sometimes i am under the impression the app is slow, then I try to open DailyGammon in browser . So I can check is it the app or the server or the internet connection who is slow.

very important to know when the app is the slow part how tp reproduce that.

**Reply by hape42** [reply]
> Okay, maybe I must be more precise what my observation and problem is :-)
> 
> Let's assume I'm in a game. I click "roll dice" and see the dice
> 
> * I click on a checker to move it. Unexpectedly much time passes, it's not immediate that the checker moves
> * I click on a second checker to move it. Again more time than I know from off-line backgammons passes
> * I see that the position is not good and want to try out a different move. I press "undo" and again some time passes until the position before my two checker moves is restored.
> 
> Sometimes the time in these 3 steps is really fast, sometimes it is slow, so my assumption is that the DG app does some remote query to the DG server happens whenever I click a checker or undo.

Okay, I think I get it.

Yes, you are right. Each move is sent to the server and the server returns the HTML page with the new position. Now I have to parse this HTML page again and rebuild and display the board completely. Each move is handled like this until the "submit" or "undo" button.

There is no game logic implemented in the app. But we would have to implement it to prevent invalid moves.

And at some point you will be satisfied with your moves. What should happen then? You could then send all moves of the checkers in a batch run. Because every moved checker is a signal to the server!

And we would still have to distinguish whether the movements of the checkers are now only for testing, or the actual train is. 

As you have already guessed, this would be a larger project that needs to be well thought out.

It is not impossible, but certainly very time consuming.

**Reply by kagsteiner** [reply]
Yes, I totally agree. We would have to develop a separate full backgammon board with all rules about legal moves, and a good synchronization with the DG board. This would be almost a rewrite of big pieces of the app...

**Comment by Smiles2go** [comment]
**Some simple(?) ideas:**

- In Top Page header also show number of matches pending / waiting for opponent. If slow to acquire, maybe add to Game Lounge instead. 

- Add button that opens  DG-web Top Page in a browser (automatically provide username password if needed). Gives easy access to Discussion and other DG functions not covered by the app.
-- Help button mostly covers this already, except auto-login

- Simple sounds as feedback for a move, dice roll, maybe a few other things

**not so simple**
- Add assist for game export into file that can be analyzed by at least the XG Mobile app (maybe also MacOS-hosted apps, when stored to iCloud). Maybe should only be available at end of game to reduce temptation to use as a cheat. 

- Reimagine chat handling to include history per-match and/or per-opponent. Allow quoting from prior messages. Have consistent presentation of just-received text if any, option to quote that text, and editing/composition that doesn't hide text being edited. Could have per-opponent history bounded to (say) 30 days or 30 messages (auto-deleted FIFO queues) into past. Slight challenge how best to handle when you have multiple matches open with same opponent.

**Reply by hape42** [reply]
Thank you for your ideas. I will work this into the RoadMap, but still have a few questions about some ideas:

> * In Top Page header also show number of matches pending / waiting for opponent. If slow to acquire, maybe add to Game Lounge instead.

Is this the total number of active matches minus the matches from topPage? Then it should not be very difficult to determine this number. This is also nothing that will be slow. The hardest part will be deciding where to put the information. Can you make a concrete suggestion? Best would be a screenshot of the iPhone and iPad with the respective position of text and number drawn in. 
> * Add button that opens  DG-web Top Page in a browser (automatically provide username password if needed). Gives easy access to Discussion and other DG functions not covered by the app.
>   -- Help button mostly covers this already, except auto-login

I do not know how to transfer the login cookie to safari. But otherwise we already have this solution through the help button. I will find out if it is possible to transfer the cookie from the app to safari. Then I can add this to the help feature. 
> * Simple sounds as feedback for a move, dice roll, maybe a few other things

Oh, sounds. A controversial topic. Shouldn't be that difficult. But it must be possible to turn it off. Personally, I prefer it quiet when playing. The most important thing is the license free sounds. I do not know myself with it. Can you get license free sound files? And what else besides move and roll do you have in mind?


> * Add assist for game export into file that can be analyzed by at least the XG Mobile app (maybe also MacOS-hosted apps, when stored to iCloud). Maybe should only be available at end of game to reduce temptation to use as a cheat.

I already have something in development there. I don't see the problem with the cheat so critically. After all, anyone who plays on the website could also export it at any time. For this purpose, the website also offers the export button. 
> * Reimagine chat handling to include history per-match and/or per-opponent. Allow quoting from prior messages. Have consistent presentation of just-received text if any, option to quote that text, and editing/composition that doesn't hide text being edited. Could have per-opponent history bounded to (say) 30 days or 30 messages (auto-deleted FIFO queues) into past. Slight challenge how best to handle when you have multiple matches open with same opponent.

You guess right. A very difficult and complex topic. The whole chat topic is very unsatisfactorily solved. But to make it better, you have to dig very deep into the app and rethink a few basic things and then develop them again.

**Reply by Smiles2go** [reply]
> > * In Top Page header also show number of matches pending / waiting for opponent.
> 
> Is this the total number of active matches minus the matches from topPage? 

I was thinking the full total. Really more a concern when thinking about how many new tournaments or matches to sign up for. 

> - The hardest part will be deciding where to put the information. Can you make a concrete suggestion?

I suggest add list-count at top of each of the new Lists pages, in whitespace above column labels. 

You are already numbering entries, but it can take a while to scroll all the way down. 
> 
> 
> >   -- Help button mostly covers this already, except auto-login
> 
> 
> But otherwise we already have this solution through the help button. I will find out if it is possible to transfer the cookie from the app to safari. Then I can add this to the help feature. 
> 
Sounds good. Thanks.

> > * Simple sounds as feedback for a move, dice roll, maybe a few other things
> 
> 
> Oh, sounds. A controversial topic. Shouldn't be that difficult. But it must be possible to turn it off. 

Ok.


> The most important thing is the license free sounds. 

> 
Yes, royalty-free sounds are key. Can record some for this if needed. 

> And what else besides move and roll do you have in mind?

Maybe also
- Message received 
- Round/game won
-  ... lost

Also, we don't want game sounds to interrupt music if playing in background. Many apps somehow have their sounds mixed on top of music. 
> 
> 
> > * Add assist for game export into file that can be analyzed by at least the XG Mobile app
> 
> I already have something in development there. I don't see the problem with the cheat so critically. After all, anyone who plays on the website could also export it at any time. For this purpose, the website also offers the export button. 

True. Still, is a bit messy, so some help would be nice. Maybe an option from the new Lists area where active and completed games can now be selected. 

> > * Reimagine chat handling to include history per-match and/or per-opponent. Allow quoting from prior messages. Have consistent presentation of just-received text if any, option to quote that text, and editing/composition that doesn't hide text being edited. Could have per-opponent history bounded to (say) 30 days or 30 messages (auto-deleted FIFO queues) into past. Slight challenge how best to handle when you have multiple matches open with same opponent.
> 
> You guess right. A very difficult and complex topic. The whole chat topic is very unsatisfactorily solved. But to make it better, you have to dig very deep into the app and rethink a few basic things and then develop them again.
> 

A start might be build a common message/chat display and compose function that provides a more flexible, uniform UI. Please allow composition (with sent-message view) in a movable window that can be repositioned / temporarily hidden, floating over board (or whatever). Also please allow temporary portrait orientation during composition so keyboard doesn't cover so much of the screen. 

Needs parameters like:
- text or handle : message source (player or system) if known
- associated opponent or game if known
- text to display, if any
- Boolean : response allowed? If True, offer to begin message composition mode, with option quote received text

With common message-handler in place, can work out message store / view / quote later to add to it.

**Reply by hape42** [reply]
> Also please allow temporary portrait orientation during composition so keyboard doesn't cover so much of the screen.

Unfortunately, individual views cannot be displayed in portrait mode. Either all, or none.
I could remove the restriction to landscape again. But then all views would have to be reworked, so that there are also meaningful views in portrait mode. I tried this about a year ago. it's an incredible amount of detail work. i discontinued it after a few days of unsuccessful design attempts. But the hardest part will be getting the board to play so that it remains somewhat readable.

If someone has a solution, I'm happy to hand over this part and learn how to implement this sensibly with portrait and landscape mode. 

Translated with www.DeepL.com/Translator (free version)

**Reply by hape42** [reply]
> Please allow composition (with sent-message view) in a movable window that can be repositioned / temporarily hidden, floating over board (or whatever). 

I don't have to allow it, I didn't prevent it or forbid it. I do not know how to implement this. If someone here has a solution I would be very happy about a support.

**Reply by hape42** [reply]
I tried to use WhatsApp in landscape mode. They have not been able to solve this either. 80% of the screen is covered by the keyboard.

If you know an app for the iphone that solved the problem in landscape mode satisfactorily, please post screenshots here. I would then contact their developers and ask for tips.

**Comment by kagsteiner** [comment]
Read an interesting discussion on the DG board about the lacking / difficult social aspect of DG; One main complaint that I totally understand was that you can enter a message to your opponent, but this is not stored and with so many games in parallel it is difficult/impossible to keep an ongoing dialog.

I have seen that we already have the solution for the app on the roadmap - "chat history", and when I thought about the problem in the discussion I found that this actually would be really great to have; e.g. if you click on your opponent during a game you can see the full chat history, and whenever you enter something as a chat text or the app presents you a chat text from the opponent it is stored locally (or even in the iCloud...?)

If you don't mind I'll extend the section in "suggestions for roadmap" with more details of what could be nice. I would love to implement this myself, but this is probably over my abilities, even with help of ChatGPT :-)

**Reply by Smiles2go** [reply]
Thanks for continuing to evolve DG app capabilities. I know it can be a lot of work. 

Agree that chat is really weak in web-DG, inherited by client app. 

Social aspect in online games is important when as a society we seem to get out of our caves less, have less IRL contact with fellow humans (and the occasional extraterrestrial visitors). With Jordan L showing a surprising interest in adding features recently, might this be something he's thinking about? Does anyone correspond with him?

Possibilities are endless and many variations are out there. Whatever we do, please DO NOT require tie back to Facebook or Twitter identities or services. 

Focus first on MVP sort of feature set and see if it is worth enhancing from there. 

Somewhere else I thought I'd suggested starting with a common message-display API and service initially just to have a common UI for presenting incoming DG messages and for composing / sending, vice the 3 or 4 different styles app currently uses. Please don't make the one true method always full-screen, unless there's an easy way to hide/recall the message and composition views. 

Can steal some ideas (I mean take inspiration) from Apple iMessage perhaps. 

With a single basic message handler then it becomes much easier to look at adding features to that single API and service, things like history (with review and search). And yes, I'd love iCloud store (as with ratings), allowing sharing between iPhone and iPad using same AppleID.

**Reply by kagsteiner** [reply]
Totally agree that we shouldn’t mess around with trying to integrate our messaging with social networks (and don’t see any reason to do this).

**Reply by Smiles2go** [reply]
Many other apps ask users to link to social platforms, maybe as advertising to get more exposure for their app. I don't know why, but I try to avoid. 

 - Jon

**Comment by kagsteiner** [comment]
I don’t yet have a concrete idea how to do this, but what would be really cool would be some support to figure out whether to chat with a certain opponent in German or English. Probably applies to any language vs English.

Situation:
I’m playing with X, whose nationality I don’t recall, and where the name doesn’t clearly signal that the person speaks German (e.g. CaptnBlaubaer is a clear hint :-) ). The chat dialog at the end of my moves pops up and I want to write something (“man that was a lot of luck for me”).

Today: I either open the web UI, go to the match, select the player, see that she/he has text written in German, or a location in Germany, and I use German, or else I use English. Inconvenient.

Tomorrow: With some magic of the DailyGammon app I can recognise in the chat box whether I can use German (maybe a collapse/expand of player info?)

**Reply by hape42** [reply]
Once I have managed to make the app usable in every direction, such features are not really a big problem. 

When the chat history and the text modules / phrases are developed, you could also program user information such as language (and perhaps even more). A flag of the language in which you usually speak with your opponent could then appear in the chat. The information could also be saved in the iCloud so that you have the information you have saved about your opponent available, regardless of the device.

**Comment by kagsteiner** [comment]
One idea that came to my mind (very very low priority :-) ): I personally struggle to understand the race situation in one glance. I'm wondering if it would be possible to visualize the pip count difference better than just, well, two numbers.

Maybe color-coding like in the attached images?
![Bildschirmfoto 2024-02-12 um 08 59 40](https://github.com/hape42/dailygammon/assets/25194620/5800686f-60bc-4d9f-99eb-cd35aa7ebf88)
![Bildschirmfoto 2024-02-12 um 09 02 22](https://github.com/hape42/dailygammon/assets/25194620/d5de6e95-96c4-482e-bad7-cc19c6a9ad27)

Maybe also just paint the text in colors or something? Or maybe a text "you are (far) ahead/behind"?
